core/src/main/java/io/atomix/core/multimap/impl/ConsistentSetMultimapService.java

```diff
@@ -1,723 +1,228 @@
+   * value, false otherwise.
    */
-  protected boolean containsEntry(Commit<? extends ContainsEntry> commit) {
-    MapEntryValue entryValue =
-        backingMap.get(commit.value().key());
-    if (entryValue == null) {
-      return false;
-    } else {
-      Match valueMatch = Match.ifValue(commit.value().value());
-      return entryValue
-          .values()
-          .stream()
-          .anyMatch(byteValue -> valueMatch.matches(byteValue));
-    }
-  }
+  @Query
+  boolean containsEntry(String key, byte[] value);
 
   /**
-   * Handles a Clear commit.
+   * If the key-value pair does not already exist adds either the key value
+   * pair or the value to the set of values associated with the key and
+   * returns true, if the key-value pair already exists then behavior is
+   * implementation specific with some implementations allowing duplicates
+   * and others ignoring put requests for existing entries.
    *
-   * @param commit Clear commit
+   * @param key   the key to add
+   * @param value the value to add
+   * @return true if the map has changed because of this call,
+   * false otherwise
    */
-  protected void clear(Commit<Void> commit) {
-    backingMap.clear();
-  }
+  @Command
+  boolean put(String key, byte[] value);
 
   /**
-   * Handles a KeySet commit.
+   * Removes the key-value pair with the specified values if it exists. In
+   * implementations that allow duplicates which matching entry will be
+   * removed is undefined.
    *
-   * @param commit KeySet commit
-   * @return a set of all keys in the multimap
+   * @param key   the key of the pair to be removed
+   * @param value the value of the pair to be removed
+   * @return true if the map changed because of this call, false otherwise.
    */
-  protected Set<String> keySet(Commit<Void> commit) {
-    return ImmutableSet.copyOf(backingMap.keySet());
-  }
+  @Command
+  boolean remove(String key, byte[] value);
 
   /**
-   * Handles a Keys commit.
+   * Removes the key-value pairs with the specified key and values if they
+   * exist. In implementations that allow duplicates each instance of a key
+   * will remove one matching entry, which one is not defined. Equivalent to
+   * repeated calls to {@code remove()} for each key value pair but more
+   * efficient.
    *
-   * @param commit Keys commit
-   * @return a multiset of keys with each key included an equal number of
-   * times to the total key-value pairs in which that key participates
+   * @param key    the key of the pair to be removed
+   * @param values the set of values to be removed
+   * @return true if the map changes because of this call, false otherwise.
    */
-  protected Multiset<String> keys(Commit<Void> commit) {
-    Multiset keys = HashMultiset.create();
-    backingMap.forEach((key, mapEntryValue) -> {
-      keys.add(key, mapEntryValue.values().size());
-    });
-    return keys;
-  }
+  @Command("removeAllValues")
```

Comment:
do we check for conflicting command annotations?

Reply:
yep